### PR TITLE
fix: unhandled promise rejection in scheduleWorkflowReminder

### LINKED ISSUE #12340: [CAL-2703] Fix Cron job scheduleEmailReminders
Cron job for scheduling workflow reminders is failing. 

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-2703](https://linear.app/calcom/issue/CAL-2703/fix-cron-job-scheduleemailreminders)</sub>

### PR DESCRIPTION
**Pull request by CarinaWolli**
## What does this PR do?

Fixes `unhandled promise rejection` error that was caused by await before Promise.allSettled. 

Now we first load all needed workflowReminders and after that we loop over all workflowReminders and create the promises. 

fixes #12340

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: packages/features/ee/workflows/api/scheduleEmailReminders.ts [hunk 1]
```diff
@@ -40,10 +70,11 @@ async function handler(req: NextApiRequest, res: NextApiResponse) {
   // delete batch_ids with already past scheduled date from scheduled_sends
   const pageSize = 90;
   let pageNumber = 0;
-  const deletePromises: Promise<any>[] = [];
+
+  const remindersToDelete: { referenceId: string | null }[] = [];
 
   while (true) {
-    const remindersToDelete = await prisma.workflowReminder.findMany({
+    const newRemindersToDelete = await prisma.workflowReminder.findMany({
```
Comment (CarinaWolli) [inline packages/features/ee/workflows/api/scheduleEmailReminders.ts:77]: This is changed everywhere below too. The while loop now exclusively gathers all workflowReminders. Once we've collected all the reminders, we process them in a separate step, generating the necessary promises to handle the required requests.

@hariombalhara That should solve the `unhandled promise rejection`, right?

Comment (hariombalhara) [inline packages/features/ee/workflows/api/scheduleEmailReminders.ts:77]:
```suggestion
    const remindersToDeletePromises = await prisma.workflowReminder.findMany({
```

Comment (hariombalhara) [inline packages/features/ee/workflows/api/scheduleEmailReminders.ts:77]: This logic to get remindersToDelete can be abstracted out in a separate function now easily as it's just fetching workflowReminders page by page till it finds no more.
`getAllRemindersToDelete` could be the name.

File: packages/features/ee/workflows/api/scheduleEmailReminders.ts [hunk 2]
```diff
@@ -94,10 +129,10 @@ async function handler(req: NextApiRequest, res: NextApiResponse) {
   //cancel reminders for cancelled/rescheduled bookings that are scheduled within the next hour
   pageNumber = 0;
 
-  const allPromisesCancelReminders: Promise<any>[] = [];
+  const remindersToCancel: { referenceId: string | null; id: number }[] = [];
 
   while (true) {
-    const remindersToCancel = await prisma.workflowReminder.findMany({
+    const newRemindersToCancel = await prisma.workflowReminder.findMany({
```
Comment (hariombalhara) [inline packages/features/ee/workflows/api/scheduleEmailReminders.ts:135]: Same here.

File: packages/features/ee/workflows/api/scheduleEmailReminders.ts [hunk 3]
```diff
@@ -153,8 +193,10 @@ async function handler(req: NextApiRequest, res: NextApiResponse) {
   pageNumber = 0;
   const sendEmailPromises: Promise<any>[] = [];
 
+  const unscheduledReminders: PartialWorkflowReminder[] = [];
+
   while (true) {
-    const unscheduledReminders = await prisma.workflowReminder.findMany({
+    const newUnscheduledReminders = await prisma.workflowReminder.findMany({
```
Comment (hariombalhara) [inline packages/features/ee/workflows/api/scheduleEmailReminders.ts:199]: Same here. It makes more sense to have a workflowReminder getter with just different filters.

File: packages/features/ee/workflows/api/scheduleEmailReminders.ts [hunk 4]
```diff
@@ -214,190 +256,183 @@ async function handler(req: NextApiRequest, res: NextApiResponse) {
...
+        const batchIdResponse = await client.request({
+          url: "/v3/mail/batch",
+          method: "POST",
+        });
+
+        const batchId = batchIdResponse[1].batch_id;
+
+        if (reminder.workflowStep.action !== WorkflowActions.EMAIL_ADDRESS) {
+          await sendEmailPromises.push(
+            sgMail.send({
```
Comment (hariombalhara) [inline packages/features/ee/workflows/api/scheduleEmailReminders.ts:389]:
```suggestion
           await sgMail.send({
```

File: packages/features/ee/workflows/api/scheduleEmailReminders.ts [hunk 5]
```diff
@@ -214,190 +256,183 @@ async function handler(req: NextApiRequest, res: NextApiResponse) {
...
+      if (emailContent.emailSubject.length > 0 && !emailBodyEmpty && sendTo) {
+        const batchIdResponse = await client.request({
+          url: "/v3/mail/batch",
+          method: "POST",
+        });
+
+        const batchId = batchIdResponse[1].batch_id;
+
+        if (reminder.workflowStep.action !== WorkflowActions.EMAIL_ADDRESS) {
+          await sendEmailPromises.push(
```
Comment (hariombalhara) [inline packages/features/ee/workflows/api/scheduleEmailReminders.ts:388]:
```suggestion
          sendEmailPromises.push(
```

File: packages/features/ee/workflows/api/scheduleEmailReminders.ts [hunk 6]
```diff
@@ -214,190 +256,183 @@ async function handler(req: NextApiRequest, res: NextApiResponse) {
...
+          },
+        });
       }
+    } catch (error) {
+      logger.error(`Error scheduling Email with error ${error}`);
     }
-    pageNumber++;
   }
 
-  Promise.allSettled(sendEmailPromises).then((results) => {
```
Comment (hariombalhara) [inline packages/features/ee/workflows/api/scheduleEmailReminders.ts:392]: Do we intentionally want to fail when there is a failure in sending even one email?

**Review by hariombalhara — CHANGES_REQUESTED** [review]
The approach  looks correct to me now @CarinaWolli. Left a few comments to make the code more DRY and comfortable to understand.

Also, I think it could make sense to unit test it as it seems like an important endpoint to me(Could be a followup)

File: packages/features/ee/workflows/api/scheduleEmailReminders.ts (hunk 6 above)
Comment (CarinaWolli) [reply packages/features/ee/workflows/api/scheduleEmailReminders.ts:392]: Not sure why I removed that 🤔 But we definitely don't want to fail when the sending of one email fails

**Comment by CarinaWolli** [comment]
@hariombalhara Thank you for the feedback. I made the adjustments to the code, can you take a look again? 

Also, I will follow up with some tests: https://github.com/calcom/cal.com/issues/12355

**Review by hariombalhara — APPROVED** [review]
LGTM now !!

**Pull request merged**